darts/models/forecasting/regression_model.py

```diff
@@ -1032,3 +1032,358 @@ def __init__(self):
+                    for lag in self.lags.get("target", [])
+                    for component in target_ts.components
+                ]
+                + [
+                    f"past_cov_{component}_lag{lag}"
+                    for lag in self.lags.get("past", [])
+                    for component in past_covs_ts.components
+                ]
+                + [
+                    f"fut_cov_{component}_lag{lag}"
+                    for lag in self.lags.get("future", [])
+                    for component in fut_covs_ts.components
+                ]
+                + (
+                    list(target_ts.static_covariates.columns)
+                    if target_ts.has_static_covariates
+                    # if isinstance(target_ts.static_covariates, pd.DataFrame)
+                    else []
+                )
+            )
+
+            indices = [
+                i
+                for i, col in enumerate(feature_list)
+                for cat in categorical_covariates
+                if cat and cat in col
+            ]
+            col_names = [feature_list[i] for i in indices]
+
+            return indices, col_names
+
+    @staticmethod
+    def _cast_float_to_int(data: np.ndarray, col_indices: List[int]) -> pd.DataFrame:
+        """
+        Casts the columns of the data array at the given indices to int and returns a pd.DataFrame.
+        """
+        df = pd.DataFrame(data)
+        if col_indices:
+            for col in col_indices:
+                df[col] = df[col].astype(int)
+        return df
+
+    def _fit_model(
+        self,
+        target_series,
+        past_covariates,
+        future_covariates,
+        max_samples_per_ts,
+        **kwargs,
+    ):
+        """
+        Custom fit function for `RegressionModelWithCategoricalCovariates` models, adding logic to let the model
+        handle categorical features directly.
+        """
+
+        training_samples, training_labels = self._create_lagged_data(
+            target_series,
+            past_covariates,
+            future_covariates,
+            max_samples_per_ts,
+        )
+
+        # if training_labels is of shape (n_samples, 1) flatten it to shape (n_samples,)
+        if len(training_labels.shape) == 2 and training_labels.shape[1] == 1:
+            training_labels = training_labels.ravel()
+
+        cat_col_indices, _ = self._get_categorical_features(
+            target_series,
+            past_covariates,
+            future_covariates,
+        )
+
+        # Updated each time the model is fit such that the indices can be easily retrieved in the predict method later
+        self._categorical_col_indices = cat_col_indices
+
+        training_samples_df = self._cast_float_to_int(training_samples, cat_col_indices)
+
+        kwargs[self.categorical_fit_param_name] = cat_col_indices
+        self.model.fit(
+            training_samples_df,
+            training_labels,
+            **kwargs,
+        )
```

Comment:
I think we should avoid for now having the conversion from numpy array to pandas DataFrame.
This can get quite slow for large arrays. 

Let's exclude Catboost categorical support for now in this PR until we come up with a robust solution. Thanks to your work, adding support at a later time will be very easy! 👍 🚀  :)

Also, now we can make use of the parent method again

```suggestion
        cat_col_indices, _ = self._get_categorical_features(
            target_series,
            past_covariates,
            future_covariates,
        )
        # Updated each time the model is fit such that the indices can be easily retrieved in the predict method later
        self._categorical_col_indices = cat_col_indices
        kwargs[self.categorical_fit_param_name] = cat_col_indices
        super()._fit_model(
            target_series=target_series,
            past_covariates=past_covariates,
            future_covariates=future_covariates,
            max_samples_per_ts=max_samples_per_ts,
            **kwargs
        )
```